agent/bootstrap.go

```diff
@@ -0,0 +1,845 @@
+package agent
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/buildkite/agent/shell"
+	"github.com/buildkite/agent/vendor/src/github.com/go-version"
+	"github.com/buildkite/agent/vendor/src/github.com/mitchellh/go-homedir"
+	"github.com/buildkite/agent/vendor/src/github.com/nightlyone/lockfile"
+)
+
+type Bootstrap struct {
+	// The command to run
+	Command string
+
+	// The ID of the job being run
+	JobID string
+
+	// If the bootstrap is in debug mode
+	Debug bool
+
+	// The repository that needs to be cloned
+	Repository string
+
+	// The commit being built
+	Commit string
+
+	// The branch of the commit
+	Branch string
+
+	// The tag of the job commit
+	Tag string
+
+	// Should git submodules be checked out
+	GitSubmodules bool
+
+	// If the commit was part of a pull request, this will container the PR number
+	PullRequest string
+
+	// The provider of the the project
+	ProjectProvider string
+
+	// Slug of the current project
+	ProjectSlug string
+
+	// Name of the agent running the bootstrap
+	AgentName string
+
+	// Should the bootstrap remove an existing checkout before running the job
+	CleanCheckout bool
+
+	// Whether or not to run the hooks/commands in a PTY
+	RunInPty bool
+
+	// Are aribtary commands allowed to be executed
+	CommandEval bool
+
+	// Path where the builds will be run
+	BuildPath string
+
+	// Path to the buildkite-agent binary
+	BinPath string
+
+	// Path to the global hooks
+	HooksPath string
+
+	// Paths to automatically upload as artifacts when the build finishes
+	AutomaticArtifactUploadPaths string
+
+	// A custom destination to upload artifacts to (i.e. s3://...)
+	ArtifactUploadDestination string
+
+	// Whether or not to automatically authorize SSH key hosts
+	AutoSSHFingerprintVerification bool
+
+	// The running environment for the bootstrap file as each task runs
+	env *shell.Environment
+
+	// Current working directory that shell commands get executed in
+	wd string
+}
+
+// Prints a line of output
+func printf(format string, v ...interface{}) {
+	fmt.Printf("%s\n", fmt.Sprintf(format, v...))
+}
+
+// Prints a bootstrap formatted header
+func headerf(format string, v ...interface{}) {
+	fmt.Printf("~~~ %s\n", fmt.Sprintf(format, v...))
+}
+
+// Prints an info statement
+func commentf(format string, v ...interface{}) {
+	fmt.Printf("\033[90m# %s\033[0m\n", fmt.Sprintf(format, v...))
+}
+
+// Shows a buildkite boostrap error
+func errorf(format string, v ...interface{}) {
+	headerf(":rotating_light: \033[31mBuildkite Error\033[0m")
+	printf(format, v...)
+}
+
+// Shows a buildkite boostrap warning
+func warningf(format string, v ...interface{}) {
+	printf("\033[33m⚠️ Buildkite Warning: %s\033[0m", fmt.Sprintf(format, v...))
+	printf("^^^ +++")
+}
+
+// Shows the error text and exits the bootstrap
+func fatalf(format string, v ...interface{}) {
```

Comment:
I would call this `exitf` just to because it doesn't show a trace like `fatalf` (which uses panic("message"))